From 7469c9dc9bb7e4502b3fe0b52c08c5411b3282a8 Mon Sep 17 00:00:00 2001
From: xichen12 <xi.a.chen@intel.com>
Date: Thu, 18 Oct 2018 16:37:08 +0800
Subject: [PATCH 0832/1214] move sdhci-pci & mmc_block to fs_initcall_sync

due to mmc_rescan will cost 140ms for mmc detect.
this will block dm-0 and rootfs setup.
thus make kernel initial longer.

so move them to early stage make mmc detect finishes
early, finally save kernel initial time.

saves about 140ms.

Change-Id: I6072b60acd3caa1337bc8b5aee26edfb7446dbb7
Tracked-On:
Signed-off-by: xichen12 <xi.a.chen@intel.com>
---
 drivers/mmc/core/block.c          |  2 +-
 drivers/mmc/host/sdhci-pci-core.c | 12 +++++++++++-
 2 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/core/block.c b/drivers/mmc/core/block.c
index 517bd4f..d75b276 100644
--- a/drivers/mmc/core/block.c
+++ b/drivers/mmc/core/block.c
@@ -3284,7 +3284,7 @@ static void __exit mmc_blk_exit(void)
 	bus_unregister(&mmc_rpmb_bus_type);
 }
 
-module_init(mmc_blk_init);
+fs_initcall_sync(mmc_blk_init);
 module_exit(mmc_blk_exit);
 
 MODULE_LICENSE("GPL");
diff --git a/drivers/mmc/host/sdhci-pci-core.c b/drivers/mmc/host/sdhci-pci-core.c
index 2d2743e..f039184 100644
--- a/drivers/mmc/host/sdhci-pci-core.c
+++ b/drivers/mmc/host/sdhci-pci-core.c
@@ -2047,7 +2047,17 @@ static struct pci_driver sdhci_driver = {
 	},
 };
 
-module_pci_driver(sdhci_driver);
+static int __init sdhci_driver_init(void)
+{
+	return pci_register_driver(&sdhci_driver);
+}
+fs_initcall_sync(sdhci_driver_init);
+
+static void __exit sdhci_driver_exit(void)
+{
+	pci_unregister_driver(&sdhci_driver);
+}
+module_exit(sdhci_driver_exit);
 
 MODULE_AUTHOR("Pierre Ossman <pierre@ossman.eu>");
 MODULE_DESCRIPTION("Secure Digital Host Controller Interface PCI driver");
-- 
2.7.4

